actionview/lib/action_view/helpers/tag_helper.rb

```diff
@@ -26,7 +26,149 @@ module TagHelper
+          else
+            content_tag_string(name, content || '', options, escape_attributes)
+          end
+        end
+
+        def content_tag_string(name, content, options, escape = true)
+          tag_options = tag_options(options, escape) if options
+          content     = ERB::Util.unwrapped_html_escape(content) if escape
+          "<#{name}#{tag_options}>#{PRE_CONTENT_STRINGS[name]}#{content}</#{name}>".html_safe
+        end
+
+        def tag_options(options, escape = true)
+          return if options.blank?
+          output = ""
+          sep    = " ".freeze
+          options.each_pair do |key, value|
+            if TAG_PREFIXES.include?(key) && value.is_a?(Hash)
+              value.each_pair do |k, v|
+                next if v.nil?
+                output << sep
+                output << prefix_tag_option(key, k, v, escape)
+              end
+            elsif BOOLEAN_ATTRIBUTES.include?(key)
+              if value
+                output << sep
+                output << boolean_tag_option(key)
+              end
+            elsif !value.nil?
+              output << sep
+              output << tag_option(key, value, escape)
+            end
+          end
+          output unless output.empty?
+        end
+
+        def boolean_tag_option(key)
+          %(#{key}="#{key}")
+        end
+
+        def tag_option(key, value, escape)
+          if value.is_a?(Array)
+            value = escape ? safe_join(value, " ".freeze) : value.join(" ".freeze)
+          else
+            value = escape ? ERB::Util.unwrapped_html_escape(value) : value
+          end
+          %(#{key}="#{value}")
+        end
+
+        private
+          def prefix_tag_option(prefix, key, value, escape)
+            key = "#{prefix}-#{key.to_s.dasherize}"
+            unless value.is_a?(String) || value.is_a?(Symbol) || value.is_a?(BigDecimal)
+              value = value.to_json
+            end
+            tag_option(key, value, escape)
+          end
+
+          def respond_to_missing?(*args)
+            true
+          end
+
+          def method_missing(called, *args, &block)
+            tag_string(called, *args, &block)
+          end
+
+      end
+
+      # Returns an HTML tag. Supports two syntax variants: legacy and modern.
+      #
+      # === Modern syntax
+      # Modern syntax uses following format:
+      #   tag.<name>(args)
+      # Returns an HTML tag. Content has to be a string. If content is passed
+      # than tag is surrounding the content. Otherwise tag will be empty. You
+      # can also use a block to pass the content inside ERB templates. Result
+      # is by default HTML5 compliant. Include +escape_attributes+: +false+ 
+      # in options to disable attribute value escaping. The tag will be
+      # generated with related closing tag unless tag represents a
+      # void[https://www.w3.org/TR/html5/syntax.html#void-elements] element.
+      #
```

Comment:
I wanted to see if interspersing the examples inbetween documentation would read nicer.

```ruby
# === Building HTML tags
# Builds HTML5 compliant tags with a tag proxy. Every tag can be built with:
#
#   tag.<tag name>(optional content, options)
#
# where tag name can be e.g. br, div, section, article, or any tag really.
#
# ==== Passing content
# Tags can pass content to embed within it:
#
#   tag.h1 'All shit fit to print' # => <h1>All shit fit to print</h1>
#
#   tag.div tag.p('Hello world!')  # => <div><p>Hello world!</p></div>
#
# Content can also be captured with a block. Great for ERB templates:
#
#   <%= tag.p do %>
#     The next great American novel starts here.
#   <% end %>
#   # => <p>The next great American novel starts here.</p>
#
# ==== Options
# Any passed options becomes attributes on the generated tag.
#
#   tag.section class: %w( kitties puppies )
#   # => <section class="kitties puppies"></section>
#
#   tag.section id: dom_id(@post)
#   # => <section id="<generated dom id>"></section>
#
# Pass true for any attributes that can render with no values like +disabled+.
#
#   tag.input type: 'text', disabled: true
#   # => <input type="text" disabled="disabled">
#
# HTML5 <tt>data-*</tt> attributes can be set with a single +data+ key
# pointing to a hash of sub-attributes.
#
# To play nicely with JavaScript conventions sub-attributes are dasherized.
#
#   tag.article data: { user_id: 123 }
#   # => <article data-user-id="123"></article>
#
# Thus <tt>data-user-id</tt> can be accessed as <tt>dataset.userId</tt>.
#
# Data attribute values are encoded to JSON, with the exception of strings, symbols and
# BigDecimals.
# This may come in handy when using jQuery's HTML5-aware <tt>.data()</tt>
# from 1.4.3.
#
#   tag.div data: { city_state: %w( Chigaco IL ) }
#   # => <div data-city-state="[&quot;Chicago&quot;,&quot;IL&quot;]"></div>
#
# The generated attributes are escaped by default, but it can be turned off with
# +escape_attributes+.
#
#   tag.img src: 'open & shut.png'
#   # => <img src="open &amp; shut.png">
#
#   tag.img src: 'open & shut.png', escape_attributes: false
#   # => <img src="open & shut.png">
#
# The tag builder respects
# [HTML5 void elements](https://www.w3.org/TR/html5/syntax.html#void-elements)
# if no content is passed, and omits closing tags for those elements.
#
#   # A standard element:
#   tag.div # => <div></div>
#
#   # A void element:
#   tag.br  # => <br>
```

What do you think, @marekkirejczyk, @rafaelfranca? 😁